optimized/utils.py

```
@@ -0,0 +1,271 @@
+        A new NumPy array representing the perturbed graph.
+    """
+    perturbed_graph = graph.copy()
+    for r_idx, c_idx, new_value in perturbation:
+        perturbed_graph[r_idx, c_idx] = new_value
+    return perturbed_graph
+
+
+def get_perturbations(
+    graph: npt.NDArray[np.int_],
+    perturbation_count: int = 1,
+    fix_incoming_to: list[int] = list(),
```

Ruff (0.12.2)

Do not use mutable data structures for argument defaults

Replace with `None`; initialize within function

(B006)